src/java/org/apache/cassandra/cql3/conditions/ColumnCondition.java

```diff
@@ -531,6 +534,25 @@ private static boolean setOrListAppliesTo(AbstractType<?> type, Iterator<Cell<?>
             return operator == Operator.EQ || operator == Operator.LTE || operator == Operator.GTE;
         }
 
+        private static boolean valueAppliesTo(Iterator<Cell<?>> iter, ByteBuffer value, boolean appliesToSetOrMapKeys)
+        {
+            while(iter.hasNext())
+            {
+                // for lists and map values we use the cell value; for sets and map keys we use the cell name
+                ByteBuffer cellValue = appliesToSetOrMapKeys ? iter.next().path().get(0) : iter.next().buffer();
+                int comparison = BytesType.instance.compare(cellValue, value);
```

Comment:
Wouldn't it be better to rely on the already existing `compareWithOperator()`?

Reply:
I'm not sure about this, it seems to me that use of the compareWithOperator method is related to the nature of the column of the condition. It's a general behavior of the SimpleColumnCondition; if the target column is a multicell collection it uses a MultiCellCollectionBound that not relies on the compareWithOperator, if not it uses a SimpleBound that relies on it.

So a "CONTAINS" clause on a frozen collection will use a SimpleBound and the compareWithOperator method but the same clause on a non frozen collection will use a MultiCellCollectionBound (and this specific compare code)

Maybe I miss something but it seems that a change at this level will have a much bigger impact on existing code (impact I tried to limit for the moment).

Reply:
Correct that's clear.

But notice how across the file comparisons are done through `type .compare()` methods and then `evaluateComparisonWithOperator()` is called. In your solution iiuc you compare ByteBuffers directly through the Bytes type and the check for `==0` side stepping the logic already existing for those needs. You don't check for the key or element type and use that API.

I think you should check key/element type to get the right type comparator, instead of going to a raw BB comparison. And then you could rely on `evaluateComparisonWithOperator()` to piggyback the already existing logic and have a single point of responsibility for that.

I would then, although this more a personal cosmetic choice, try to make it pretty by consolidating everything in an overloaded `compareWithOperator()`. But this is more involved refactoring etc.

Reading again the short comment I dropped I have to apologize as it was really misleading lol. I was thinking along the lines of jumping back into the already existing logics, my bad.

Reply:
Thanks for your answer you clear some confusion and I see your point now :). I made some change to be consistant with a regular `type .compare()` instead of the raw BB comparison. 

However I don't feel confortable to fallback on the `evaluateComparisonWithOperator()` because current call are performed `when comparison != 0`. It's fine for the existing operators but in this specific case of contains operators it seems pretty straight forward to check for `==0` because we don't want to iterate over the whole collection values (we will if the condition is not satisfied but otherwise as soon as match is found the condition is satisfied).

Reply:
Hi,

can I ask you to please don't squash on each push? Then I can see what changed exactly from commit to commit. I look at a few PRs at the same time and my memory is not that good to remember the whole code lol.

I see you now use the right type comparator depending on the collection type. I still would call `evaluateComparisonWithOperator()` with an `EQ` operator on each of the iterator's loop. Iiuc that would be the same behavior we have atm. Let's see what other committer thinks when he reviews this.

Reply:
All instances where the comparison `!= 0` in the class call `evaluateComparisonWithOperator()`. So for consistency and in case the logic in that method changed (add logging, update a metric, God knows,...) I would use it:

```
int res = type.compare(cellValue, value);
return res == 0 ? 0 : evaluateComparisonWithOperator(res, EQ);
```

Reply:
If we fallback on `evaluateComparisonWithOperator()` when a comparison is `!=0` we will return **false** for the condition and it can be wrong (other item of the collection can satisfied the contains condition).

Exemple =>  evaluation of the `l CONTAINS 'bar'` condition against the collection `l = ['foo', 'bar', 'foobar']`

The comparison 'foo' vs 'bar' will be `!= 0` but the condition still applies to the collection. We have to perform a second comparison 'bar' vs 'bar' `==0` to return the final status of the condition (**true**).

Reply:
Oh right sorry. It's been a long time :-) Sthg like that to bring `evaluateComparisonWithOperator()` into the picture or sthg similar is what I am trying to suggest.

```
 boolean res = false;
 while(iter.hasNext())
        {
            // for lists and map values we use the cell value; for sets and map keys we use the cell name
            ByteBuffer cellValue = appliesToSetOrMapKeys ? iter.next().path().get(0) : iter.next().buffer();
            int comp = type.compare(cellValue, value);
            if(comp == 0)
                return true;
            else
                res = evaluateComparisonWithOperator(comp, EQ);
        }
        return res;
```

Reply:
`evaluateComparisonWithOperator` is simply there to avoid duplicating code. Adding some side logic to it likes logging or metric will lead to some problems no matter what, considering the code calling it.
For me the current approach makes more sense.